drivers/input/input_stmpe811.c

```diff
@@ -0,0 +1,490 @@
+		/* Calculate positions values */
+		uldata_xy = (data_xy[0] << 16) | (data_xy[1] << 8) | data_xy[2];
+		data->touch_x = (uldata_xy >> 12U) & BIT_MASK(12);
+		data->touch_y = uldata_xy & BIT_MASK(12);
+
+		/* Reset FIFO */
+		ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG, FIFO_STA_CLEAR);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG,
+					    FIFO_STA_OPERATIONAL);
+		if (ret < 0) {
+			return ret;
+		}
+	} else if (touch_detected == 0) {
+		data->touch_detected = 0;
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int stmpe811_ts_enable_it(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+	uint8_t tmp;
+
+	/* Get the current value of the INT_EN register */
+	int ret = i2c_reg_read_byte_dt(&config->bus, STMPE811_INT_EN_REG, &tmp);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Enable FIFO above threshold interrupt */
+	tmp |= STMPE811_GIT_FTH;
+
+	ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_INT_EN_REG, tmp);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Get the current value of the interrupt control register  */
+	ret = i2c_reg_read_byte_dt(&config->bus, STMPE811_INT_CTRL_REG, &tmp);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Enable global interrupts */
+	tmp |= STMPE811_GIT_EN;
+
+	return i2c_reg_write_byte_dt(&config->bus, STMPE811_INT_CTRL_REG, tmp);
+}
+
+static int stmpe811_clear_global_it(const struct device *dev, uint8_t source)
+{
+	const struct stmpe811_config *config = dev->config;
+
+	/* Write 1 to the bits that have to be cleared */
+	return i2c_reg_write_byte_dt(&config->bus, STMPE811_INT_STA_REG, source);
+}
+
+static int stmpe811_process(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+	struct stmpe811_data *data = dev->data;
+
+	int err = stmpe811_ts_get_state(dev);
+
+	if (err) {
+		return err;
+	}
+
+	if (data->touch_detected) {
+		int x = config->screen_width - (data->touch_x / 4096.0) * config->screen_width;
+		int y = config->screen_height - (data->touch_y / 4096.0) * config->screen_height;
+
+		input_report_abs(data->dev, INPUT_ABS_X, x, false, K_FOREVER);
+		input_report_abs(data->dev, INPUT_ABS_Y, y, false, K_FOREVER);
+		input_report_key(data->dev, INPUT_BTN_TOUCH, 1, true, K_FOREVER);
+	}
+
+	/* Start polling to detect release */
+	data->poll_release_counter = 0;
+	k_sem_give(&data->poll_lock);
+
+	return 0;
+}
+
+static void stmpe811_work_handler(struct k_work *work)
+{
+	struct stmpe811_data *data = CONTAINER_OF(work, struct stmpe811_data, work);
+
+	stmpe811_process(data->dev);
+}
+
+static void poll_release(const struct device *dev)
+{
+	struct stmpe811_data *data = dev->data;
+	const struct stmpe811_config *config = dev->config;
+
+	int err;
+	uint32_t current_cycles;
+	uint32_t cycles_diff;
+	uint32_t wait_period;
+
+	k_usleep(data->poll_period_us);
+	k_timepoint_t poll_time_end = sys_timepoint_calc(K_USEC(data->poll_timeout_us));
+
+	while (true) {
```

Comment:
hey one higher level comment before getting into details: you sure you want a thread here? feels like you could get away with a `k_work_delayable` and save the system resources by having one less thread just for the touchscreen. I can tell you based this on the npcx keyboard scan driver (that can get a bit messy with the matrix scanning code) but I don't think it's really needed in this case and you could get away but just firing the content of this loop on the sysworkq at the specified delay without messing too much with cycles compensation etc...

Reply:
Thanks for the suggestion! We changed it to use `k_work_delayable`.